src/typecheck/arith.ml

```diff
@@ -2,6 +2,95 @@
 module Id = Dolmen.Std.Id
 module Term = Dolmen.Std.Term
 
+(* Ae arithmetic *)
+(* ************************************************************************ *)
+module Ae = struct
+
+  module Tff
+      (Type : Tff_intf.S)
+      (Ty : Dolmen.Intf.Ty.Ae_Arith with type t := Type.Ty.t)
+      (T : Dolmen.Intf.Term.Ae_Arith with type t := Type.T.t
+                                        and type ty := Type.Ty.t) = struct
+
+      type _ Type.err +=
+      | Expected_arith_type : Type.Ty.t -> Term.t Type.err
+      | Incompatible_Arith_Types : Type.Ty.t * Type.Ty.t -> Term.t Type.err
+
+      let dispatch1 env (mk_int, mk_real) ast t =
+        let ty = T.ty t in
+        match Ty.view ty with
+        | `Int -> mk_int t
+        | `Real -> mk_real t
+        | _ -> Type._error env (Ast ast) (Expected_arith_type ty)
+
+      let dispatch2 ?(strict = true) env (mk_int, mk_real) ast a b =
+        let tya = T.ty a in
+        let tyb = T.ty b in
+        match Ty.view tya, Ty.view tyb with
+        | `Int, `Int -> mk_int a b
+        | `Real, `Real -> mk_real a b
+        | (`Real, `Int | `Int, `Real) when not strict -> mk_real a b
```

Comment:
in this case, you should add explicit type coercions on the ``` `Int``` argument, before giving it to the mk_real function

Reply:
According to Alt-Ergo's doc [doc](https://ocamlpro.github.io/alt-ergo/Input_file_formats/Native/02_types/02_01_builtin.html#built-in-types) we should only do that when the ```**.``` is used. But I was wondering if there is a reason why we would not want to do it for the other operators? aka: whenever one of the two operands is a real the other one (if it's an int) is coerced to a real and the mk_real function is applied.

Reply:
Well, it can be useful to do that, but it can also silently turn a lot of int computation into real computations silently if you don't pay attention, so it's a kind of double-edged sword. In any case, it's up to alt-ergo to define what its language specifies regarding that, ^^

Reply:
Ok